### PR TITLE
[CUDA] Add opt-in SM90 DeepGEMM for MatMulBlockQuantizedFp8Weight

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description

Add an opt-in SM90 DeepGEMM path for `MatMulBlockQuantizedFp8Weight` when `a_scale` is present and `block_size=128`. Set `ORT_FP8_MATMUL_DEEPGEMM=1` before session creation to use native FP8 multiplication with the existing per-row/per-128-K weight scales, avoiding the FP16/BF16 weight-dequantization buffer.

Dispatch preserves the existing small-M GEMV and requires M<=128, N>=2048, K divisible by 128, N divisible by 64, and at least 8 MiB of FP8 weights. Larger batches and smaller weights retain the existing path because benchmarks found regressions outside this initial range. The flag defaults to off.

The implementation prepares scales on the compute stream each call, clears the upstream kernel's accumulating FP32 output, tiles output scratch within the existing scratch limit, and converts/adds bias in the existing rounding order. It supports changing scales and CUDA graph replay. Native FP8 changes intermediate operand rounding, so deployment still requires model-accuracy validation.

### Motivation and Context

Reuse the DeepGEMM dependency added in #32122 to reduce W8A8 prefill latency. This uses its dense SM90 1D1D kernel to match the operator's scale layout.

H200, CUDA 13.0.48, same Release binary with flag off/on; BF16, W8A8, bias, median complete-operator GPU latency over 50 CUDA graph replays (16 calls per replay, after 10 warmups):

| M | N | K | Flag off | Flag on | Speedup |
|---:|---:|---:|---:|---:|---:|
| 64 | 4096 | 4096 | 27.78 us | 17.23 us | 1.61x |
| 128 | 4096 | 4096 | 33.37 us | 19.77 us | 1.69x |
| 64 | 11008 | 4096 | 68.73 us | 31.73 us | 2.17x |
| 128 | 16384 | 5120 | 122.65 us | 63.76 us | 1.92x |

FP16 showed similar gains. Nsight Systems confirmed native `sm90_fp8_gemm_1d1d_impl` dispatch and the replacement of weight dequantization plus cuBLAS. Timings include activation quantization, scale preparation, output clearing, GEMM, conversion, and bias; they exclude host dispatch and session creation. Full measurements, methodology, and reproduction commands are in `docs/contrib_ops/cuda/matmul_block_scaled_fp8_experiments.md`.

Validation:

- Built the CUDA provider and Python binding on Linux/H200; also compiled the operator source without `USE_DEEP_GEMM`.
- All 44 main benchmark records and 32 additional small/fallback records passed accuracy checks.
- Six new integration tests passed: FP16/BF16, independent scales, scale changes/zero scale, clipping, bias, tile/pipeline boundaries, bounded scratch, fallback cases, nondefault stream, CUDA graph replay, and explicit native-dispatch verification. Run with `ORT_TEST_FP8_DEEPGEMM=1 python onnxruntime/test/python/contrib_ops/test_matmul_fp8_deepgemm.py` against the built package.
- Compute Sanitizer memcheck: tile/pipeline-boundary and scratch-tiling tests passed with zero errors.
- clang-format, ruff check/format, and `git diff --check` passed for changed files.

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/math/matmul_block_scaled_fp8.cc [hunk 1]
```diff
@@ -127,6 +128,44 @@ Status MatMulBlockQuantizedFp8Weight::ComputeImpl(OpKernelContext* context) cons
                         (k_i % 16 == 0) && (block_size_ % 16 == 0);
   const bool fuse_act_qdq = use_gemv && !FusedFp8ActivationQdqDisabled();
 
+#if defined(USE_DEEP_GEMM)
+  // Native FP8 changes intermediate rounding, so it is opt-in and requires the model's
+  // activation scale. Keep the tuned small-M path and unsupported layouts on the fallback.
+  if (enable_deep_gemm_ && !use_gemv && a_scale != nullptr && block_size_ == 128 &&
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/math/matmul_block_scaled_fp8.cc:134]: Fixed in b209998f5b. The schema SetDoc and a_scale description now distinguish default FP16/BF16 dequantization from the opt-in SM90 native FP8 path, including the intermediate-rounding difference and fallback behavior. Updated the class comment and regenerated this operator's section in ContribOperators.md. The local lightweight-protobuf build strips runtime docstrings, so generation used source docstrings with rebuilt schema metadata; the same procedure reproduced the previous section exactly. The CUDA/Python rebuild, all six focused integration tests, and repository-pinned lint checks passed.

File: onnxruntime/test/python/contrib_ops/profile_matmul_block_scaled.py [hunk 2]
```diff
@@ -40,9 +40,9 @@
 import numpy as np
 import torch
 from onnx import TensorProto, helper
+from onnxruntime.capi.onnxruntime_pybind11_state import Fail as OrtFail
 
 import onnxruntime
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/test/python/contrib_ops/profile_matmul_block_scaled.py:44]: Fixed in b209998f5b using the repository-pinned lintrunner/ruff versions. The import block now passes both RUFF and RUFF-FORMAT checks.

**Review by baijumeswani — APPROVED** [review]
This is in good shape. The new DeepGEMM path is opt-in, limited to exact SM90 devices and supported shapes, and keeps the existing GEMV and dequantize-plus-cuBLAS paths as fallbacks. The scale packing, output clearing, tiled scratch use, output stride handling, and bias rounding order match the kernel contracts and existing operator behavior.

The test coverage is strong: it checks both data types, changing scales, zero scale, clipping, bias, tile boundaries, scratch tiling, CUDA graph replay, fallback cases, and confirms native dispatch instead of accepting a fallback-only pass. I also built this head for SM80 on A100 and ran all 15 focused MatMulBlockQuantizedFp8Weight tests with ORT_FP8_MATMUL_DEEPGEMM=0 and 1; both runs passed, so the opt-in does not change non-SM90 behavior.
